File: sdks/node/src/cli-secrets.ts
import fsp from 'node:fs/promises';
import readline from 'node:readline';
import { Writable } from 'node:stream';

const MAX_SECRET_FILE_BYTES = 64 * 1024;

interface SecretSources {
  readonly privateKeyFile?: string;
  readonly tokenFile?: string;
}

interface SecretTerminal {
  readonly interactive: boolean;
  readHidden(prompt: string): Promise<string>;
}

export interface InstallSecrets {
  readonly privateKey: string;
  readonly token?: string;
}

function parseSingleLineSecret(raw: string, label: string, allowEmpty: boolean): string {
  const value = raw.endsWith('\r\n')
    ? raw.slice(0, -2)
    : raw.endsWith('\n')
      ? raw.slice(0, -1)
      : raw;

  if (!allowEmpty && value.length === 0) {
    throw new Error(`${label} is empty`);
  }
  if (value.includes('\r') || value.includes('\n') || value.includes('\0')) {
    throw new Error(`${label} must contain exactly one line`);
  }
  return value;
}

async function readSecretFile(filePath: string, label: string): Promise<string> {
  const stats = await fsp.lstat(filePath);
  if (!stats.isFile()) {
    throw new Error(`${label} file is not a regular file: ${filePath}`);
  }
  if (stats.size > MAX_SECRET_FILE_BYTES) {
    throw new Error(`${label} file exceeds ${MAX_SECRET_FILE_BYTES} bytes: ${filePath}`);
  }
  if (process.platform !== 'win32' && (stats.mode & 0o077) !== 0) {
    throw new Error(`${label} file must be accessible only by its owner: ${filePath}`);
  }

  const raw = await fsp.readFile(filePath, 'utf-8');
  return parseSingleLineSecret(raw, label, false);
}

async function readHidden(prompt: string): Promise<string> {
  if (!process.stdin.isTTY || !process.stderr.isTTY) {
    throw new Error('hidden input requires an interactive terminal');
  }

  let muted = false;
  const output = new Writable({
    write(chunk, encoding, callback) {
      if (!muted) process.stderr.write(chunk, encoding);
      callback();
    },
  });
  const terminal = readline.createInterface({
    input: process.stdin,
    output,
    terminal: true,
    historySize: 0,
  });

  process.stderr.write(prompt);
  muted = true;
  try {
    return await new Promise<string>((resolve, reject) => {
      terminal.once('SIGINT', () => reject(new Error('secret input cancelled')));
      terminal.question('', resolve);
    });
  } finally {
    muted = false;
    terminal.close();
    process.stderr.write('\n');
  }
}

const defaultTerminal: SecretTerminal = {
  interactive: Boolean(process.stdin.isTTY && process.stderr.isTTY),
  readHidden,
};

export async function resolveInstallSecrets(
  sources: SecretSources,
  terminal: SecretTerminal = defaultTerminal,
): Promise<InstallSecrets> {
  let privateKey: string;
  if (sources.privateKeyFile) {
    privateKey = await readSecretFile(sources.privateKeyFile, 'private key');
  } else if (terminal.interactive) {
    privateKey = parseSingleLineSecret(
      await terminal.readHidden('Private key: '),
      'private key',
      false,
    );
  } else {
    throw new Error(
      'private key input requires an interactive terminal or --private_key_file <path>',
    );
  }

  let token: string | undefined;
  if (sources.tokenFile) {
    token = await readSecretFile(sources.tokenFile, 'API token');
  } else if (terminal.interactive) {
    token = parseSingleLineSecret(
      await terminal.readHidden('API token (optional): '),
      'API token',
      true,
    ) || undefined;
  }

  return { privateKey, token };
}

File: sdks/go/cmd/elydora/cli_secrets.go
package main

import (
	"crypto/ed25519"
	"encoding/base64"
	"errors"
	"fmt"
	"io"
	"os"
	"runtime"
	"strings"
	"unicode/utf8"
)

const maxSecretFileBytes = 64 * 1024

type installSecretSources struct {
	privateKeyFile string
	tokenFile      string
}

type installSecrets struct {
	privateKey string
	token      string
}

type secretTerminal interface {
	Interactive() bool
	ReadHidden(prompt string) (string, error)
}

type standardSecretTerminal struct {
	input  *os.File
	output *os.File
}

func newStandardSecretTerminal() standardSecretTerminal {
	return standardSecretTerminal{input: os.Stdin, output: os.Stderr}
}

func (terminal standardSecretTerminal) Interactive() bool {
	return isInteractiveTerminal(terminal.input) &&
		isInteractiveTerminal(terminal.output)
}

func (terminal standardSecretTerminal) ReadHidden(prompt string) (string, error) {
	if !terminal.Interactive() {
		return "", errors.New("hidden input requires an interactive terminal")
	}
	if _, err := fmt.Fprint(terminal.output, prompt); err != nil {
		return "", fmt.Errorf("write secret prompt: %w", err)
	}
	raw, readErr := readHiddenLine(terminal.input)
	_, newlineErr := fmt.Fprintln(terminal.output)
	if readErr != nil {
		clear(raw)
		return "", errors.Join(
			fmt.Errorf("read hidden input: %w", readErr),
			newlineErr,
		)
	}
	if newlineErr != nil {
		clear(raw)
		return "", fmt.Errorf("finish secret prompt: %w", newlineErr)
	}
	value := string(raw)
	clear(raw)
	return value, nil
}

func parseSingleLineSecret(raw, label string, allowEmpty bool) (string, error) {
	if len(raw) > maxSecretFileBytes {
		return "", fmt.Errorf("%s exceeds %d bytes", label, maxSecretFileBytes)
	}
	if !utf8.ValidString(raw) {
		return "", fmt.Errorf("%s must contain UTF-8 text", label)
	}
	value := raw
	if strings.HasSuffix(value, "\r\n") {
		value = strings.TrimSuffix(value, "\r\n")
	} else if strings.HasSuffix(value, "\n") {
		value = strings.TrimSuffix(value, "\n")
	}
	if !allowEmpty && value == "" {
		return "", fmt.Errorf("%s is empty", label)
	}
	if strings.ContainsAny(value, "\r\n\x00") {
		return "", fmt.Errorf("%s must contain exactly one line", label)
	}
	return value, nil
}

func validateSecretFile(info os.FileInfo, path, label string) error {
	if !info.Mode().IsRegular() {
		return fmt.Errorf("%s file is not a regular file: %s", label, path)
	}
	if info.Size() > maxSecretFileBytes {
		return fmt.Errorf("%s file exceeds %d bytes: %s", label, maxSecretFileBytes, path)
	}
	if runtime.GOOS != "windows" && info.Mode().Perm()&0077 != 0 {
		return fmt.Errorf("%s file must be accessible only by its owner: %s", label, path)
	}
	return nil
}

func readSecretFile(path, label string) (string, error) {
	return readSecretFileWithOpen(path, label, os.Open)
}

func readSecretFileWithOpen(
	path string,
	label string,
	openFile func(string) (*os.File, error),
) (value string, returnErr error) {
	before, err := os.Lstat(path)
	if err != nil {
		return "", fmt.Errorf("inspect %s file at %s: %w", label, path, err)
	}
	if err := validateSecretFile(before, path, label); err != nil {
		return "", err
	}
	if !os.SameFile(before, before) {
		return "", fmt.Errorf("establish %s file identity at %s", label, path)
	}

	file, err := openFile(path)
	if err != nil {
		return "", fmt.Errorf("open %s file at %s: %w", label, path, err)
	}
	defer func() {
		if closeErr := file.Close(); closeErr != nil {
			returnErr = errors.Join(
				returnErr,
				fmt.Errorf("close %s file at %s: %w", label, path, closeErr),
			)
		}
	}()

	after, err := file.Stat()
	if err != nil {
		return "", fmt.Errorf("inspect open %s file at %s: %w", label, path, err)
	}
	if err := validateSecretFile(after, path, label); err != nil {
		return "", err
	}
	if !os.SameFile(before, after) {
		return "", fmt.Errorf("%s file changed while opening: %s", label, path)
	}

	raw, err := io.ReadAll(io.LimitReader(file, maxSecretFileBytes+1))
	if err != nil {
		return "", fmt.Errorf("read %s file at %s: %w", label, path, err)
	}
	defer clear(raw)
	if len(raw) > maxSecretFileBytes {
		return "", fmt.Errorf("%s file exceeds %d bytes: %s", label, maxSecretFileBytes, path)
	}
	return parseSingleLineSecret(string(raw), label, false)
}

func resolveInstallSecrets(
	sources installSecretSources,
	terminal secretTerminal,
) (installSecrets, error) {
	var privateKey string
	var err error
	if sources.privateKeyFile != "" {
		privateKey, err = readSecretFile(sources.privateKeyFile, "private key")
	} else if terminal.Interactive() {
		privateKey, err = terminal.ReadHidden("Private key: ")
		if err == nil {
			privateKey, err = parseSingleLineSecret(privateKey, "private key", false)
		}
	} else {
		return installSecrets{}, errors.New(
			"private key input requires an interactive terminal or --private-key-file <path>",
		)
	}
	if err != nil {
		return installSecrets{}, err
	}

	var token string
	if sources.tokenFile != "" {
		token, err = readSecretFile(sources.tokenFile, "API token")
	} else if terminal.Interactive() {
		token, err = terminal.ReadHidden("API token (optional): ")
		if err == nil {
			token, err = parseSingleLineSecret(token, "API token", true)
		}
	}
	if err != nil {
		return installSecrets{}, err
	}
	return installSecrets{privateKey: privateKey, token: token}, nil
}

func validatePrivateKey(privateKey string) error {
	seed, err := base64.RawURLEncoding.DecodeString(privateKey)
	if err != nil {
		return fmt.Errorf("decode private key: %w", err)
	}
	defer clear(seed)
	if len(seed) != ed25519.SeedSize {
		return fmt.Errorf(
			"private key seed must be %d bytes, got %d",
			ed25519.SeedSize,
			len(seed),
		)
	}
	return nil
}

func rejectLegacySecretArguments(arguments []string) error {
	legacyOptions := map[string]string{
		"--private-key": "--private-key-file",
		"-private-key":  "--private-key-file",
		"--token":       "--token-file",
		"-token":        "--token-file",
	}
	for _, argument := range arguments {
		option := strings.SplitN(argument, "=", 2)[0]
		if replacement, exists := legacyOptions[option]; exists {
			return fmt.Errorf(
				"%s exposes credentials in process arguments; use %s or hidden terminal input",
				option,
				replacement,
			)
		}
	}
	return nil
}
